src/configuration/configuration_manager.cpp

```diff
@@ -0,0 +1,155 @@
+  str << "Socket Family: " << ConfigurationUtil::GET_STRING(ConfigurationId::socket_family) << std::endl;
+  str << "Statistics: " << (ConfigurationUtil::GET_INT(ConfigurationId::stats_mode) ? "enabled" : "disabled") << std::endl;
+  str << "Max Connections: " << ConfigurationUtil::GET_INT(ConfigurationId::max_connections) << std::endl;
+  str << "Index Tuner: " << (ConfigurationUtil::GET_BOOL(ConfigurationId::index_tuner) ? "enabled" : "disabled") << std::endl;
+  str << "Layout Tuner: " << (ConfigurationUtil::GET_BOOL(ConfigurationId::layout_tuner) ? "enabled" : "disabled") << std::endl;
+  str << "Code-generation: " << (ConfigurationUtil::GET_BOOL(ConfigurationId::codegen) ? "enabled" : "disabled") << std::endl;
+
+  str << std::endl;
+  str << "//===---------------------------------------------------===//" << std::endl;
+
+  return str.str();
+}
+
+void ConfigurationManager::ShowInfo() {
+  LOG_INFO("%s", GetInfo().c_str());
+}
+
+void ConfigurationManager::DefineConfig(
+    ConfigurationId id, const std::string &name,
+    void* gflags_ptr, const std::string &description,
+    const type::Value &value, const type::Value &default_value,
+    bool is_mutable, bool is_persistent) {
+  if (config.count(id) > 0) {
+    throw ConfigurationException("configuration " + name + " already exists");
+  }
+  config[id] = Param(name, gflags_ptr, description,
+                     value, default_value,
+                     is_mutable, is_persistent);
+}
+
+type::Value ConfigurationManager::GetValue(ConfigurationId id) {
+  // TODO: Look up the value from catalog
+  // Because querying a catalog table needs to create a new transaction and
+  // creating transaction needs to get config values,
+  // it will be a infinite recursion here.
+
+  auto param = config.find(id);
+  return param->second.value;
+}
+
+void ConfigurationManager::SetValue(ConfigurationId id, const type::Value &value) {
+  auto param = config.find(id);
+  param->second.value = value;
+  switch (value.GetTypeId()) {
+    case type::TypeId::INTEGER:
+      *(int32_t*)param->second.gflags_ptr = value.GetAs<int32_t>();
+      break;
+    case type::TypeId::BOOLEAN:
+      *(bool*)param->second.gflags_ptr = value.GetAs<bool>();
+      break;
+    case type::TypeId::VARCHAR:
+      *(std::string*)param->second.gflags_ptr = value.GetAs<const char*>();
+      break;
+    default:
+      throw new ConfigurationException("unsupported type");
+  }
+  if (catalog_initialized) {
+    insert_into_catalog(param->second);
+  }
+}
+
+void ConfigurationManager::insert_into_catalog(const Param &param) {
+  auto config_catalog = catalog::SettingsCatalog::GetInstance();
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+  type::AbstractPool *pool = pool_.get();
+  config_catalog->DeleteSetting(param.name, txn);
+  config_catalog->InsertSetting(param.name, param.value.ToString(),
```

Comment:
Could you also leave a comment why you are not updating the catalog, but delete-then-inserting?